pkg/extensions/tbtc/tbtc.go

```diff
@@ -0,0 +1,269 @@
+	logger.Infof("initializing tbtc extensions")
+
+	manager := &extensionsManager{handle}
+
+	err := manager.initializeRetrievePubkeyExtension(ctx, 150*time.Minute)
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize retrieve pubkey extension: [%v]",
+			err,
+		)
+	}
+
+	logger.Infof("tbtc extensions have been initialized")
+
+	return nil
+}
+
+type depositEventHandler func(deposit string)
+
+type watchDepositEventFn func(
+	handler depositEventHandler,
+) (subscription.EventSubscription, error)
+
+type watchKeepClosedFn func(deposit string) (
+	keepClosedChan chan struct{},
+	keepClosedCancel func(),
+	err error,
+)
+
+type submitDepositTxFn func(deposit string) error
+
+type extensionsManager struct {
+	handle Handle
+}
+
+func (em *extensionsManager) initializeRetrievePubkeyExtension(
+	ctx context.Context,
+	timeout time.Duration,
+) error {
+	logger.Infof("initializing retrieve pubkey extension")
+
+	startEventSubscription, err := em.monitorAndAct(
+		"retrieve pubkey",
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return em.handle.OnDepositCreated(handler)
+		},
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return em.handle.OnDepositRegisteredPubkey(handler)
+		},
+		em.watchKeepClosed,
+		func(deposit string) error {
+			return em.handle.RetrieveSignerPubkey(deposit)
+		},
+		timeout,
+	)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		<-ctx.Done()
+		startEventSubscription.Unsubscribe()
+		logger.Infof("retrieve pubkey extension has been disabled")
+	}()
+
+	logger.Infof("retrieve pubkey extension has been initialized")
+
+	return nil
+}
+
+// TODO:
+//  1. Filter incoming events by operator interest.
+//  2. Incoming events deduplication.
+//  3. Resume extensions executions after client restart.
+func (em *extensionsManager) monitorAndAct(
+	extensionName string,
+	monitoringStartFn watchDepositEventFn,
+	monitoringStopFn watchDepositEventFn,
+	keepClosedFn watchKeepClosedFn,
+	actFn submitDepositTxFn,
+	timeout time.Duration,
+) (subscription.EventSubscription, error) {
+	handleStartEvent := func(deposit string) {
+		logger.Infof(
+			"triggering [%v] extension execution for deposit [%v]",
+			extensionName,
+			deposit,
+		)
+
+		stopEventChan := make(chan struct{})
+
+		stopEventSubscription, err := monitoringStopFn(
+			func(stopEventDeposit string) {
+				if deposit == stopEventDeposit {
+					stopEventChan <- struct{}{}
+				}
+			},
+		)
+		if err != nil {
+			logger.Errorf(
+				"could not setup stop event handler for [%v] "+
+					"extension execution and deposit [%v]: [%v]",
+				extensionName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer stopEventSubscription.Unsubscribe()
+
+		keepClosedChan, keepClosedSubscriptionCancel, err := keepClosedFn(deposit)
+		if err != nil {
+			logger.Errorf(
+				"could not setup keep closed handler for [%v] "+
+					"extension execution and deposit [%v]: [%v]",
```

Comment:
`could not setup keep closed handler for [%v] monitoring for deposit [%v]` perhaps?